parser/transactioncache/transactioncache.go

```diff
@@ -0,0 +1,151 @@
+package transactioncache
+
+import (
+	"fmt"
+
+	"github.com/dave/dst"
+)
+
+// TransactionCache is responsible for tracking existing transactions within a Go application.
+// It maintains the following components:
+//
+//   - Transactions: A map where each key is a transaction name, and the value is a list of expressions
+//     that are active within the lifespan of the transaction.
+//
+//   - Functions: A map that stores already seen functions alongside their declarations. This is useful
+//     for tracking transactions that span multiple function calls.
+//
+//   - TransactionState: A map that tracks whether a transaction is open or has ended, ensuring that
+//     no further expressions are added once a transaction is marked as ended.
+type TransactionCache struct {
+	Transactions     map[*dst.Ident][]dst.Expr
+	Functions        map[string]*dst.FuncDecl
+	TransactionState map[*dst.Ident]bool // Track whether a transaction is closed
+}
+
+func NewTransactionCache() *TransactionCache {
+	return &TransactionCache{
+		Transactions:     make(map[*dst.Ident][]dst.Expr),
+		Functions:        make(map[string]*dst.FuncDecl),
+		TransactionState: make(map[*dst.Ident]bool),
+	}
+}
+
+// AddCall adds an expression to the list of expressions associated with a transaction.
+// It first checks if the transaction is closed, and if so, it does not add the expression.
+// If the expression is an 'End' call directly on the transaction, it marks the transaction as closed.
+// If the 'End' call is part of a segment [ex: defer txn.StartSegment.End()], it does not mark the transaction as closed.
+func (tc *TransactionCache) AddCall(transaction *dst.Ident, expr dst.Expr) {
+	if tc.Transactions == nil {
+		tc.Transactions = make(map[*dst.Ident][]dst.Expr)
+	}
+	// Check if the transaction is closed
+	if closed, exists := tc.TransactionState[transaction]; exists && closed {
+		return // Do not add calls to a closed transaction
+	}
+	// Check if the call is an End method directly on the transaction
+	if callExpr, ok := expr.(*dst.CallExpr); ok {
+		if selExpr, ok := callExpr.Fun.(*dst.SelectorExpr); ok {
+			if selExpr.Sel.Name == "End" {
+				// Check if the End method is called directly on the transaction
+				if ident, ok := selExpr.X.(*dst.Ident); ok && ident.Name == transaction.Name {
+					tc.TransactionState[transaction] = true // Mark transaction as closed
+				} else if selExpr.X.(*dst.CallExpr) != nil {
+					// This is likely part of a segment operation, do not mark transaction as closed
+					return
```

Comment:
Do we not want to add the expression to the list in this case? This return results in no expression getting appended to the list of expr stmts for the transaction.